File: 257.binary-tree-paths.py
#
# @lc app=leetcode id=257 lang=python3
#
# [257] Binary Tree Paths
#
import copy
# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.res = []
        self.subpath = []

    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if not root:
            return []
        if root:
            self.subpath.append(root.val)
        self.helper(root)
        return self.res
        

    def helper(self,root):
        if not root.left and not root.right:
            self.res.append("->".join([str(i) for i in self.subpath]))
            return 
        if root.left:
            self.subpath.append(root.left.val)
            self.helper(root.left)
            self.subpath.pop()
        if root.right:
            self.subpath.append(root.right.val)
            self.helper(root.right)
            self.subpath.pop()
        
        
# @lc code=end

File: golang/257.binary-tree-paths.go
/*
 * @lc app=leetcode.cn id=257 lang=golang
 *
 * [257] Binary Tree Paths
 */
package main

import (
	"fmt"
	"strings"
)

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func binaryTreePaths(root *TreeNode) []string {
	ret := []string{}
	collect := func(items []int) {
		itemStr := []string{}
		for _, item := range items {
			itemStr = append(itemStr, fmt.Sprintf("%d", item))
		}
		ret = append(ret, strings.Join(itemStr, "->"))
	}

	cur := []int{}

	binaryTreePathsHelper(root, cur, collect)
	return ret

}
func binaryTreePathsHelper(root *TreeNode, cur []int, collect func([]int)) {
	cur = append(cur, root.Val)

	if root.Left == nil && root.Right == nil {
		collect(cur)
		return
	}

	if root.Left != nil {
		binaryTreePathsHelper(root.Left, cur, collect)
	}
	if root.Right != nil {
		binaryTreePathsHelper(root.Right, cur, collect)
	}
}

// @lc code=end
